File: 01-lexorank/code/lexorank.py
# Lexorank: string-based ranks in a small alphabet.
# Order is lexicographic. Insert between "a" and "m" by taking midpoint -> "g".
# Rebalance is rare: we have 36^n possible ranks; only when we can't get a midpoint.

import string

# Typical alphabet: 0-9 and a-z (36 chars). Jira uses similar.
ALPHABET = string.digits + string.ascii_lowercase


def mid_string(prev: str, next: str) -> str:
    """
    Lexicographic midpoint between prev and next.
    With 36^n possible ranks, we run out of room rarely; only then rebalance.
    """
    n = max(len(prev), len(next))
    prev = prev.ljust(n, "0")
    next = next.ljust(n, ALPHABET[-1])
    result = []
    for i in range(n):
        p = ALPHABET.index(prev[i]) if i < len(prev) else 0
        q = ALPHABET.index(next[i]) if i < len(next) else len(ALPHABET) - 1
        mid = (p + q) // 2
        result.append(ALPHABET[mid])
    return "".join(result).rstrip("0") or ALPHABET[0]


def rebalance_ranks(ranks: list[str], alphabet: str = ALPHABET) -> list[str]:
    """
    Respace a list of ranks evenly in the alphabet. Used rarely when
    we've exhausted midpoints in a region (much rarer than with sparse integers).
    """
    k = len(ranks)
    if k <= 1:
        return ranks
    # Simple rebalance: assign ranks spread across the alphabet (e.g. "a", "m", "z" for 3)
    step = len(alphabet) // (k + 1)
    return [alphabet[1 + i * step] for i in range(k)]


# Example: ranks "a", "m", "z". Insert between "a" and "m" -> "g"
print(mid_string("a", "m"))  # -> "g"
# Rebalance needed only when mid_string returns None — rare in practice.

File: 01-lexorank/code/lexorank.ts
/**
 * Lexorank — string-based ordering for lists (e.g. drag-and-drop).
 * Ranks are strings in a small alphabet; order is lexicographic.
 * Insert between two ranks by taking the lexicographic midpoint (e.g. "a"–"m" → "g").
 * Rebalance is rare: we have alphabetSize^n possible ranks.
 *
 * See: 01-lexorank tutorial (sparse integers → Lexorank when gap runs out).
 */

// Typical alphabet: 0–9 and a–z (36 chars). Jira and others use similar.
const ALPHABET = "0123456789abcdefghijklmnopqrstuvwxyz";

/**
 * Lexicographic midpoint between prev and next rank.
 * Pads shorter string so we can compute position between them.
 * Returns a new rank strictly between prev and next (when they differ).
 */
function midString(prev: string, next: string, alphabet: string = ALPHABET): string {
  const n = Math.max(prev.length, next.length);
  const prevPadded = prev.padEnd(n, alphabet[0]);
  const lastChar = alphabet[alphabet.length - 1];
  const nextPadded = next.padEnd(n, lastChar);

  const result: string[] = [];
  for (let i = 0; i < n; i++) {
    const p = alphabet.indexOf(prevPadded[i]);
    const q = alphabet.indexOf(nextPadded[i]);
    const mid = Math.floor((p + q) / 2);
    result.push(alphabet[mid]);
  }
  const raw = result.join("").replace(new RegExp(`${alphabet[0]}+$`), "");
  return raw || alphabet[0];
}

/**
 * Respace a list of ranks evenly in the alphabet.
 * Use rarely when midString can't find a new rank (e.g. prev and next adjacent).
 */
function rebalanceRanks(ranks: string[], alphabet: string = ALPHABET): string[] {
  const k = ranks.length;
  if (k <= 1) return [...ranks];
  const step = Math.floor(alphabet.length / (k + 1));
  return Array.from({ length: k }, (_, i) => alphabet[1 + i * step]);
}

// ---------------------------------------------------------------------------
// Demo: core logic
// ---------------------------------------------------------------------------
function demo() {
  console.log("Lexorank — core logic\n");

  // Insert between "a" and "m" → "g"
  const betweenAM = midString("a", "m");
  console.log('midString("a", "m") =>', JSON.stringify(betweenAM));

  // Initial list: "a", "m", "z". Insert at start: between "" and "a"
  const atStart = midString("", "a");
  console.log('midString("", "a") =>', JSON.stringify(atStart));

  // Insert after last: between "z" and "zz" (use a high sentinel in practice)
  const afterZ = midString("z", "zz");
  console.log('midString("z", "zz") =>', JSON.stringify(afterZ));

  // Rebalance when we have too many ranks in a tight range
  const crowded = ["a", "b", "c", "d"];
  const respaced = rebalanceRanks(crowded);
  console.log("\nrebalanceRanks([\"a\", \"b\", \"c\", \"d\"]) =>", respaced);
}

demo();

export { ALPHABET, midString, rebalanceRanks };
